### PR TITLE
C# extractor: Improve performance by changing the caching

### PR DESCRIPTION
**Pull request by calumgrant**
Achieves a ~10% single-threaded speedup (77s -> 69s) by changing the caching to entity->label, instead of entity->id & id->label. This removes one cache lookup in the case of a cache hit, and bypasses the ID cache which can be a little slow.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 1]
```diff
@@ -50,20 +50,30 @@ public SemanticModel Model(SyntaxNode node)
...
+
+            var id = entity.Id;
+
             if (id == null)
                 throw new InternalError("Attempt to create a null entity for {0}", entity.GetType());
 
-            Label existingLabel;
-            if (labelCache.TryGetValue(id, out existingLabel))
+            // Look up the Id in the idLabelCache
+            if (idLabelCache.TryGetValue(id, out label))
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs]: Why is `idLabelCache` needed? It appears to me that this is dead code, as `entityLabelCache.TryGetValue()` returns `false` iff `idLabelCache.TryGetValue()` returns `false`?

Comment (hvitved) [reply csharp/extractor/Semmle.Extraction/Context.cs]: After thinking about this, I see that this is indeed not dead code, but can be hit when two different entities share the same ID, and therefor label. Could you put a comment saying something like that here?

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 2]
```diff
@@ -43,28 +43,81 @@ public SemanticModel Model(SyntaxNode node)
...
-                entity.Label = existingLabel;
-                return true;
+                var entity = factory.Create(this, type);
+
+                if(entityLabelCache.TryGetValue(entity, out var label))
+                {
+                    entity.Label = label;
+                    return entity;
+                }
+                else
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs]: redundant

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 3]
```diff
@@ -120,7 +154,9 @@ public void AddFreshLabel(IEntity entity)
             entity.Label = label;
         }
 
-        readonly Dictionary<IId, Label> labelCache = new Dictionary<IId, Label>();
+        readonly Dictionary<IId, ICachedEntity> idLabelCache = new Dictionary<IId, ICachedEntity>();
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs]: Add `#if DEBUG_LABELS`

File: csharp/extractor/Semmle.Extraction/Entity.cs [hunk 4]
```diff
@@ -96,48 +98,36 @@ public interface ICachedEntityFactory<Initializer, Entity> where Entity : ICache
     public static class ICachedEntityFactoryExtensions
     {
         public static Entity CreateEntity<Entity, T1, T2>(this ICachedEntityFactory<(T1, T2), Entity> factory, Context cx, T1 t1, T2 t2)
-            where Entity : ICachedEntity
-        {
-            return factory.CreateEntity(cx, (t1, t2));
-        }
+            where Entity : ICachedEntity => factory.CreateEntity2(cx, (t1, t2));
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Entity.cs]: Why not `factory.CreateEntity` -- `(t1, t2)` is never `null`? Same below

File: csharp/extractor/Semmle.Extraction/Entity.cs [hunk 5]
```diff
@@ -79,6 +79,8 @@ public interface ICachedEntity : IEntity
         void Populate();
 
         bool NeedsPopulation { get; }
+
+        object UnderlyingObject { get; }
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Entity.cs]: why added?

File: csharp/extractor/Semmle.Extraction/Symbol.cs [hunk 6]
```diff
@@ -30,6 +30,8 @@ public Initializer symbol
             get; private set;
         }
 
+        object ICachedEntity.UnderlyingObject => symbol;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Symbol.cs]: why added?

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 7]
```diff
@@ -43,28 +43,81 @@ public SemanticModel Model(SyntaxNode node)
...
+                }
+                else
+                {
+                    var id = entity.Id;
+                    label = new Label(NewId());
+                    entity.Label = label;
+                    entityLabelCache[entity] = label;
+                    DefineLabel(label, id);
+                    if (entity.NeedsPopulation)
+                        Populate(null, entity);
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs]: Why always `null` argument, and not `init as ISymbol`?

Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Context.cs]: Looks like a bug....

File: csharp/extractor/Semmle.Extraction/Entity.cs (hunk 5 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Entity.cs]: It was quite useful for debugging. I'm minded to keep it in if that's ok?

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 8]
```diff
@@ -43,28 +43,81 @@ public SemanticModel Model(SyntaxNode node)
...
-            if (labelCache.TryGetValue(id, out existingLabel))
+        /// <summary>
+        /// Creates a new entity using the factory.
+        /// Uses a different cache to <see cref="CreateEntity{Type, Entity}(ICachedEntityFactory{Type, Entity}, Type)"/>,
+        /// and can store null values.
+        /// </summary>
+        /// <param name="factory">The entity factory.</param>
+        /// <param name="init">The initializer for the entity.</param>
+        /// <returns>The new/existing entity.</returns>
+        public Entity CreateEntity2<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type type) where Entity : ICachedEntity
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Context.cs]: That's not quite the right name for it.  It's also used when you want to cache using the entity as the key, e.g.

0. When `init` is null,
1. When `Type` is a tuple. In each case the tuple may be different, but you want to index only `Item1` (e.g. `GetAlreadyCreated`)
2. When you want to use the same `init` to create different entities depending on `Type`. E.g. when you want either a `Type` or a `TypeRef`.

File: csharp/extractor/Semmle.Extraction/Context.cs (hunk 2 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Context.cs]: I prefer branches myself -- I moved the `return` out of the branches instead.

File: csharp/extractor/Semmle.Extraction/Entity.cs (hunk 4 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Entity.cs]: Following on from the comment above - you want to use `Tuple.Item1` as the key (the other columns may be different), which is the behaviour of `CreateEntity2`.

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 9]
```diff
@@ -43,28 +43,81 @@ public SemanticModel Model(SyntaxNode node)
...
+        }
 
-            entity.Label = new Label(NewId());
-            DefineLabel(entity.Label, id);
-            labelCache[id] = entity.Label;
-            return false;
+        private Entity CreateNonNullEntity<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type init) where Entity : ICachedEntity
+        {
+            if (objectEntityCache.TryGetValue(init, out var cached))
+                return (Entity)cached;
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction/Context.cs]: Ah, I failed to see that `Entity` was a type parameter; misread it as `IEntity`...

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging d73b28efe4d781112e73c3d528fbb6a57585beaf into a6003533a49a6ffd67cc123dc9ea822bcb1f5601 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-cb6b051a804c830fe4caae58321086432d72ea2c)

**new alerts:**

* 1 for Redundant ToString\(\) call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**